client/lib/wp/sync-handler/index.js

```diff
@@ -0,0 +1,200 @@
+/**
+ * Module dependencies
+ */
+import config from 'config';
+import localforage from 'localforage';
+import Hashes from 'jshashes';
+import { isWhitelisted } from './whitelist-handler';
+
+import debugFactory from 'debug';
+const debug = debugFactory( 'calypso:sync-handler' );
+
+// defaults
+const defaults = config( 'sync-handler-defaults' );
+defaults.driver = localforage[ defaults.driver ];
+
+/**
+ * SyncHandler class
+ */
+export class SyncHandler {
+	/**
+	 * Create a SyncHandler instance
+	 *
+	 * @param {Object} [setup] - localforage sync setup
+	 * @param {Function} handler - wpcom handler
+	 *
+	 * @return {Function} sync-handler wrapper
+	 */
+	constructor( setup, handler ) {
+		if ( typeof setup === 'function' ) {
+			handler = setup;
+			setup = {};
+		}
+
+		this.setup = Object.assign( {}, defaults, setup );
+		this.reqHandler = handler;
+		return this.syncHandlerWrapper( handler );
+	}
+
+	syncHandlerWrapper( handler ) {
+		return ( params, callback ) => {
+			// detect and no-sync proxy connection request
+			if ( params.metaAPI && params.metaAPI.accessAllUsersBlogs ) {
+				debug( 'skip - non-sync -proxy-handler request detected' );
+				return this.reqHandler( params, callback );
+			}
+
+			// create a copy of the request params
+			const reqParams = Object.assign( {}, params );
+			const { path } = reqParams;
+
+			// whitelist barrier
+			if ( ! isWhitelisted( params ) ) {
+				debug( 'not whitelisted: skip %o request', path );
+				return handler( params, callback );
+			};
+
+			// generate an unique resource key
+			const key = this.generateKey( reqParams );
+
+			debug( 'starting to get resource ...' );
+
+			/**
+			 * Send response immediately to the client
+			 * getting the data locally (localforage)
+			 *
+			 * @param {Object} localRecord - response stored locally
+			 */
+			const localResponseHandler = localRecord => {
+				// let's be optimistic
+				if ( localRecord ) {
+					debug( '%o stored(%o). Let\'s be optimistic ...\n', path, localRecord );
```

Comment:
Maybe update the debug statement so we state that this is the first callback run to mirror the second debug in networkFetch.